### PR TITLE
IL-5220: Adds column freezing to data-grid.

### PR DESCRIPTION
**Pull request by vinayakpatil**

### CODE REVIEW

File: src/Header/index.jsx [hunk 1]
```diff
@@ -116,6 +117,12 @@ module.exports = React.createClass({
         var cols = props.virtualColumnRendering && props.endColIndex !== null ?
                     props.columns.slice(props.startColIndex, props.endColIndex + 1) :
                     props.columns
+
+        if (props.fixedColumns.length) {
+            _.remove(cols, function(column) { return column.fixed; });
+            cols = props.fixedColumns.concat(cols);
+        }
```
Comment (gregtalarico) [inline src/Header/index.jsx]: why do we need to remove the fixed columns only to add them right back?

File: src/Row/index.jsx [hunk 2]
```diff
@@ -41,6 +42,11 @@ module.exports = React.createClass({
         var props = this.prepareProps(this.props)
         var cols = props.virtualColumnRendering && props.endColIndex !== null ? props.columns.slice(props.startColIndex, props.endColIndex + 1) : props.columns
 
+        if (props.fixedColumns.length) {
+            _.remove(cols, function(column) { return column.fixed; });
+            cols = props.fixedColumns.concat(cols);
+        }
```
Comment (gregtalarico) [inline src/Row/index.jsx]: same as above

File: src/Wrapper/index.jsx [hunk 3]
```diff
@@ -110,6 +111,14 @@ module.exports = React.createClass({
...
         var height = this.state.height !== 0 && this.state.height;
         var rowClass, cellClass, cellWidth, rowHeight, offset;
+        var cols = this.props.virtualColumnRendering && this.props.endColIndex !== null ?
+                    this.props.columns.slice(this.props.startColIndex, this.props.endColIndex + 1) :
+                    this.props.columns;
+
+        if (this.props.fixedColumns.length) {
+            _.remove(cols, function(column) { return column.fixed; });
+            cols = this.props.fixedColumns.concat(cols);
+        }
```
Comment (gregtalarico) [inline src/Wrapper/index.jsx]: same

File: src/index.jsx [hunk 4]
```diff
@@ -213,7 +214,7 @@ module.exports = React.createClass({
             var startOffset = scrollLeft;
 
             // get start column index
-            for (var i = 0; i < props.columns.length; i++){
+            for (var i = props.fixedColumns.length; i < props.columns.length; i++){
```
Comment (gregtalarico) [inline src/index.jsx:218]: do we need a check that the prop exists first?

File: src/index.jsx [hunk 5]
```diff
@@ -281,7 +282,7 @@ module.exports = React.createClass({
         var endOffset = props.totalColumnWidth - (props.style.width + state.scrollLeft);
 
         // get end column index
-        for (var i = props.columns.length - 1; i >= 0; i--){
+        for (var i = props.columns.length - 1; i >= props.fixedColumns.length; i--){
```
Comment (gregtalarico) [inline src/index.jsx:286]: same do we need to check for the prop's existence first?

File: src/index.jsx [hunk 6]
```diff
@@ -315,6 +320,37 @@ module.exports = React.createClass({
...
+            if (visibleColumnsWidth && props.style.width) {
+                width = props.style.width - visibleColumnsWidth;
+            }
+
+            width = width < 0 ? 0 : width;
+
+            if (width === 0) {
+                props.startColIndex += 1;
+            } else {
+                props.columns[props.startColIndex].sizeStyle = {
```
Comment (gregtalarico) [inline src/index.jsx:342]: should we just be writing to width and minwidth instead of a new prop?

File: src/Wrapper/index.jsx [hunk 7]
```diff
@@ -123,17 +132,19 @@ module.exports = React.createClass({
...
                     if ( j === 0 ) {
                         cellClass += ' z-first';
                     }
-                    if ( j === this.props.columns.length - 1 ) {
+                    if ( j === cols.length - 1 ) {
                         cellClass += ' z-last';
                     }
 
-                    cellWidth = this.props.columns[j].width ? {width: this.props.columns[j].width, minWidth: this.props.columns[j].width} : {minWidth: this.props.columns[j].minWidth, WebkitFlex: 1, msFlex: 1, flex: 1};
+                    cellWidth = cols[j].sizeStyle.width ? {width: cols[j].sizeStyle.width, minWidth: cols[j].sizeStyle.width}
```
Comment (gregtalarico) [inline src/Wrapper/index.jsx]: do we need the extra sizeStyle object?

File: src/index.jsx [hunk 8]
```diff
@@ -1141,6 +1186,7 @@ module.exports = React.createClass({
         }, this)
 
         this.prepareColumnSizes(props, state)
+        this.prepareFixedColumns(props, state);
```
Comment (gregtalarico) [inline src/index.jsx]: should we check for the fixed props before doing preparing the fixed cols?

File: src/index.jsx [hunk 9]
```diff
@@ -315,6 +320,37 @@ module.exports = React.createClass({
...
+        var width = 0;
+        var counter;
+
+        if (props.columns.length === props.endColIndex + 1) {
+            for (counter = props.startColIndex + 1; counter <= props.endColIndex; counter++) {
+                visibleColumnsWidth += props.columns[counter].width;
+            }
+
+            if (visibleColumnsWidth && props.style.width) {
+                width = props.style.width - visibleColumnsWidth;
```
Comment (gregtalarico) [inline src/index.jsx:334]: will visibleColumnsWidth ever be smaller that props.style.width

**Review by gregtalarico — CHANGES_REQUESTED** [review]
does this generally rely on each column having a defined width?

File: src/index.jsx (hunk 4 above)
Comment (vinayakpatil) [reply src/index.jsx:218]: It will have a default value of an empty array

File: src/index.jsx (hunk 5 above)
Comment (vinayakpatil) [reply src/index.jsx:286]: same

File: src/Header/index.jsx (hunk 1 above)
Comment (vinayakpatil) [reply src/Header/index.jsx]: This was added to support non-consecutive column freezing by moving them to the beginning. I have removed this now.

File: src/Wrapper/index.jsx (hunk 7 above)
Comment (vinayakpatil) [reply src/Wrapper/index.jsx]: sizeStyle is a built in prop used to assign inline styles to a column cell.

File: src/index.jsx (hunk 6 above)
Comment (vinayakpatil) [reply src/index.jsx:342]: In order to avoid fluctuations, I am just changing inline styles of a column cell.

File: src/index.jsx (hunk 8 above)
Comment (vinayakpatil) [reply src/index.jsx]: I have moved this into prepareColumns function to save an extra cycle.

**Comment by vinayakpatil** [comment]
I have added `minWidth` fallback in virtual column rendering logic now.

**Comment by vinayakpatil** [comment]
@talarico43 Added suggested changes, can you please review?

**Review by gregtalarico — APPROVED** [review]

**Pull request merged**